crates/pathfinder/src/rpc/api.rs

```diff
@@ -91,24 +93,34 @@ impl RpcApi {
         }
     }
 
+    pub fn with_pending_data(self, pending_data: Arc<PendingData>) -> Self {
+        Self {
+            pending_data: Some(pending_data),
+            ..self
+        }
+    }
+
+    async fn pending_data(&self) -> anyhow::Result<Arc<PendingData>> {
+        self.pending_data
+            .as_ref()
+            .cloned()
+            .ok_or_else(|| anyhow::anyhow!("Pending data not supported in this configuration"))
+    }
+
     /// Get block information given the block hash.
     pub async fn get_block_by_hash(
         &self,
         block_hash: BlockHashOrTag,
         requested_scope: Option<BlockResponseScope>,
     ) -> RpcResult<Block> {
         let block_id = match block_hash {
-            BlockHashOrTag::Tag(Tag::Pending) => {
-                let block = self
-                    .sequencer
-                    .block(block_hash.into())
-                    .await
-                    .map_err(internal_server_error)?;
-
-                let scope = requested_scope.unwrap_or_default();
-
-                return Ok(Block::from_sequencer_scoped(block, scope));
-            }
+            BlockHashOrTag::Tag(Tag::Pending) => match self.pending_data().await?.block().await {
+                Some(block) => {
+                    let scope = requested_scope.unwrap_or_default();
+                    return Ok(Block::from_sequencer_scoped(block.into(), scope));
+                }
+                None => todo!("Return no pending data error"),
```

Comment:
```suggestion
                None => {
                    let block = self.sequencer.block(block_hash.into()).await.map_err(internal_server_error)?;
                    
                    let scope = requested_scope.unwrap_or_default();
                    
                    return Ok(Block::from_sequencer_scoped(block, scope));
                }
```

Reply:
We should no longer be forwarding pending at all.

Reply:
Oki then I would move the None case over to `pending_data().await` with it returning a `RpcResult<PendingData>` which could be returned to the user right away.

Reply:
If there is a suitable channel (?) for that, then we can consider that option instead. Note that the next pending may never come though, which means this query would need a reasonable timeout from our end.

I actually wanted to use a watch channel instead of arc nonsense, but one which doesn't block the writer end. Couldn't find a suitable one though.

Reply:
> As I understood it, we serve pending data if we have it, otherwise we return some error indicating that there is no pending data currently.

I can see simple solutions here right now, unless SW are willing to provide a specific error code or agree for an empty json object `{}`:
1. just throw `INVALID_BLOCK_*` (as of latest spec it's called: [`INVALID_BLOCK_ID`](https://github.com/starkware-libs/starknet-specs/blob/3d3b2d7ad6899f64043c0deaa8a40d3d8c9b1788/api/starknet_api_openrpc.json#L1486))
2. hold on until some pending data is available and return it or ultimately timeout.

Reply:
StarkWare made the suggestion to simply use `latest` when `pending` is unavailable, which I like.